xdsl/dialects/arith.py

```diff
@@ -961,15 +961,25 @@ class Minnumf(FloatingPointLikeBinaryOp):
 class IndexCastOp(IRDLOperation):
     name = "arith.index_cast"
 
-    input: Operand = operand_def()
+    input: Operand = operand_def(IntegerType | IndexType)
 
-    result: OpResult = result_def()
+    result: OpResult = result_def(IntegerType | IndexType)
 
     traits = frozenset([Pure()])
 
     def __init__(self, input_arg: SSAValue | Operation, target_type: Attribute):
         return super().__init__(operands=[input_arg], result_types=[target_type])
 
+    def verify_(self) -> None:
+        if (
+            isinstance(self.input.type, IndexType)
+            + isinstance(self.result.type, IndexType)
+            != 1
```

Comment:
Meganit: I'd say that != each other is a bit kore idiomatic

Reply:
Hmm, how would we do that such that it works when `i32` and `i16` are used?

Reply:
I guess it would then be `type(self.input.type) != type(self.result.type)`... Hmmm. Yeah, I can see that!

Reply:
Oh I meant the isinstances, since you're checking that exactly one of the isinstance checks returns True. I thought one of the two had to be `index`, does it handle `i32` to `i16` casting, also?

Reply:
Oh sorry, the above comment was misleading. The only allowed mode is exactly one of input or output being index.

I meant that checking for `input.type != output.type` would not work, as one would need to check the type of the ssa value type^^

But it should be possible (and shorter)

Reply:
```
xdsl/dialects/arith.py:974:12: E721 Do not compare types, use `isinstance()`
```


But `if isinstance(self.input.type, type(self.result.type)):` is just less readable imo